### ISSUE TITLE
outbound shipping (FBA) example

### ISSUE THREAD

**Issue by florianbepunkt** [beardon/mws-api#67]
any chance to get an example hwo to send a outbound shipping (fba) feed with items? I really struggle with the code so far and can't find a full example.

EDITED for clarity

**Comment by florianbepunkt** [comment]
so far this is what I have

```
function testOutbound(order) {
    return this.mws.FulfillmentOutboundShipment.CreateFulfillmentOrder({
      SellerFulfillmentOrderId: order.id,
      ShippingSpeedCategory: 'Standard',
      DisplayableOrderId: order.id,
      DisplayableOrderDateTime: moment(order.invoiceDate).format(),
      DisplayableOrderComment: '',
      FulfillmentPolicy: 'FillOrKill',
      DestName: order.shippingAddress.name,
      DestAddressLine1: order.shippingAddress.address1,
      DestAddressLine2: order.shippingAddress.address2,
      DestCity: order.shippingAddress.city,
      DestPostalCode: order.shippingAddress.postalcode,
      DestCountryCode: order.shippingAddress.countryCode ,
      LineItems: [{
        DisplayableComment: '',
        GiftMessage: '',
        Quantity: 1,
        SellerFulfillmentOrderItemId: 'abcdefg',
        SellerSKU: 'CM01-FBA'
      }]
    })
  }
```

this gives me the following  `Error: The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.` which is rather strange since I use this lib successfully in the same project with the same credentials to get orders.

I'm not sure if I set the paramters right, especially when it comes to the line items (guees work on my side). Also I know that coming from the official Amazon PHP library that you need to set the SellerId as a paramter when you create a OutboundShipping request.

**Comment by tyscorp** [comment]
I have never actually used and have no way to test the FBA parts of this library because Amazon doesn't have a testing environment for MWS.

I'm guessing the signature error is a misdirect and the real error is how LineItems is being handled. The library I based this off has a weird way of representing it and I [never really fixed it](https://github.com/beardon/mws-api/blob/564d6ec8924750fd1c1f78efc956f75823ed9b62/lib/request.js#L35) because it has never been needed by my employer.

I don't think I'll really have any time soon to look at this. For now, you could debug this issue by monkey-patching the request function like so:
```
var request = mws.request;
mws.request = function (req, q, meta) {
    // modify `q` here to add LineItems
    q['Items.member.1.DisplayableComment'] = '';
    q['Items.member.1.GiftMessage'] = '';
    q['Items.member.1.Quantity'] = 1;
    q['Items.member.1.SellerFulfillmentOrderItemId'] = 'abcdefg';
    q['Items.member.1.SellerSKU'] = 'CM01-FBA';
    return request(req, q, meta);
};
```

If you wanted to submit a proper fix I'd be happy to review/merge it.

**Comment by florianbepunkt** [comment]
Okay, I tried to kind of re-engineer the repo you based this on to figure out the original intention… IMHO this whole parameter renaming thing coming from the base repo is a bit messy and I can't see the reason behind it, but getting rid of it would mean a lot of refactoring.

The way you refactored the code makes the code a lot cleaner and works for all use cases except the lists (`Type.COMPLEX`) since it messes with the list renaming/reordering that needs to take place. Do you use `Type.COMPLEX` for anything else than for the FBA? If not I can supply a fix by moving the list creation in the `set(paramName, value)` method of request.js

**Comment by tyscorp** [comment]
`Type.COMPLEX` is only used for FBA. I'd be happy if it was replaced with something good.

**Comment by florianbepunkt** [comment]
@tyscorp just submitted a PR #68 that fixes FBA outbound shipping. I haven't touched inbound shipping or inventory so can't say if they work

**Comment by gholevas** [comment]
Did PR #68 really fix this outbound shipping problem for you with the same format for LineItems you provided in the example above? I'm still getting "Error: The request signature we calculated does not match the signature you provided."

It's strange because your PR actually did fix the problem for GetFulfillmentPreview but not for CreateFulfillmentOrder

**Comment by florianbepunkt** [comment]
@gholevas that's strange indeed. I did some initial tests and everything worked but we haven't used this in production yet (but we will in about two months). But looking at the PR I basically just fixed the request so this should work for CreateFulfillmentOrder.

On the other hand there were several commits regarding fba after I created the PR which was never merged. So maybe my PR is no longer compatible.

**Comment by gholevas** [comment]
Thanks for getting back to me. The problem was actually the DisplayableOrderComment parameter was not being properly escaped and thus the signature was generated incorrectly

**Comment by eli8levit** [comment]
@gholevas @florianbepunkt Can you help please, I have exactly the same problem. I need to implement Creating Fulfillment today, but nothing works out, I get "The request signature we calculated does not match the signature you provided." 
Thank you!

**Comment by florianbepunkt** [comment]
@JediYozh Mh… sounds like you have a problem with authentication and not with the fulfillment. Have you tried Amazin MWS scratchpad… does it work? If so, what I would do to debug is use the scratchpad and compare the request with the request from mws-api… you might dig into the request module to log the actual request that is made. Or you may try to use as few parameters when creating a fulfillment order as possible (stick to mws api guide to see which parameters are necessary), then add your additional params one by one until you find what's causing the invalid checksum.

here's my code…  keep in mind that I forked the repo and implemented a PR, so not sure if this would work exactly the same with this repo.

```
import moment from 'moment';
import MWSClient from 'mws-api';


class Fulfillment {

  /**
   * [constructor description]
   * @method constructor
   * @return {[type]}    [description]
   */
  constructor() {
    this.mws = new MWSClient({
      host: process.env.AMAZON_HOST,
      accessKeyId: process.env.AMAZON_ACCESS_KEY,
      secretAccessKey: process.env.AMAZON_SECRET,
      merchantId: process.env.AMAZON_MERCHANT_ID,
      meta: {
        retry: true, // retry requests when throttled
        next: true, // auto-paginate
        limit: Infinity // only get this number of items (NOT the same as MaxRequestsPerPage)
      }
    });
  }

  /**
   * [createFulfillmentOrder description]
   * @method createFulfillmentOrder
   * @param  {[type]}               order [description]
   * @return {[type]}                     [description]
   */
  createFulfillmentOrder(order) {
    return this.mws.FulfillmentOutboundShipment.CreateFulfillmentOrder({
      SellerFulfillmentOrderId: order.id,
      ShippingSpeedCategory: 'Standard',
      DisplayableOrderId: order.id,
      DisplayableOrderDateTime: moment(order.invoiceDate).format(),
      DisplayableOrderComment: 'Vielen Dank für Ihre Bestellung.',
      FulfillmentPolicy: 'FillOrKill',
      FulfillmentAction: 'Hold',
      DestName: order.shippingAddress.name,
      DestAddressLine1: order.shippingAddress.company,
      DestAddressLine2: order.shippingAddress.address1,
      DestAddressLine3: order.shippingAddress.address2,
      DestCity: order.shippingAddress.city,
      DestPostalCode: order.shippingAddress.postalcode,
      DestCountryCode: order.shippingAddress.countryCode,
      LineItems: this.formatLineItems(order.orderItems)
    });
  }

  /**
   * [formatLineItems description]
   * @method formatLineItems
   * @param  {[type]}        orderItems [description]
   * @return {[type]}                   [description]
   */
  formatLineItems(orderItems) {
    let lineItems = [];

    for (let orderItem of orderItems) {
      lineItems.push({
        DisplayableComment: orderItem.product.title,
        GiftMessage: '',
        Quantity: orderItem.quantity,
        SellerFulfillmentOrderItemId: orderItem.product.ean,
        SellerSKU: orderItem.product.amazonSellerSKU,
      })
    }

    return lineItems;
  }
}

export default new Fulfillment();

```

**Comment by eli8levit** [comment]
@florianbepunkt it's not authentication  problem because other requests such as "**ListOrders" , "ListOrderItems", "GetOrder**" and others work great! This trouble happens only in this request. I think the reason is LineItems. @gholevas said it was "**DisplayableOrderComment**" but what did he do?
Florian, where do I get **MWSAuthToken**? Because in this module it's not required...
thank you very much for help!!

**Comment by florianbepunkt** [comment]
I agreee. If the other requests work, than the problem is the FBA part.

Can you show me your code (strip out the credentials)? Regarding the token: Have you tried to leave it just empty?

**Comment by JediYozh** [comment]
`const MWSClient = require('mws-api');
const mws = new MWSClient({
    MarketplaceId: '',  
    accessKeyId: '',
    secretAccessKey: '',
    merchantId: '',
    meta: {
        retry: true, // retry requests when throttled 
        next: true, // auto-paginate 
        limit: Infinity 
     }
  })
`
`amazonUK.FulfillmentOutboundShipment.CreateFulfillmentOrder({
    SellerFulfillmentOrderId : '1112dsd1314sdd1516',
    DisplayableOrderId : '1112dsd1314sdd1516',
    ShippingSpeedCategory: 'Standard',    
    DisplayableOrderDateTime :  "2017-09-12T08:03:17.255Z",
    DestName : "2A Stanlake Rd Shepherd's Bush, London W12 7HP",
    DestCountryCode : 'UK',
    DisplayableOrderComment: 'Your Package',
    LineItems : [{
        Quantity : 1,
        SellerFulfillmentOrderItemId: 'sdsd5d5sd5',
        SellerSKU : "2sdd2d2"
    }]
        }).then(({result}) => {
            var xml = result;
            parseString(xml, (err, result) => {
                if (err) {
                    console.log(err);
                }
                console.log(result);
            })
        })
            .catch((err) => {
                console.log(err);
            });`

**Comment by eli8levit** [comment]
`const MWSClient = require('mws-api');
const mws = new MWSClient({
    MarketplaceId: '',  
    accessKeyId: '',
    secretAccessKey: '',
    merchantId: '',
    meta: {
        retry: true, // retry requests when throttled 
        next: true, // auto-paginate 
        limit: Infinity 
    }
  })

amazonUK.FulfillmentOutboundShipment.CreateFulfillmentOrder({
    SellerFulfillmentOrderId : '1112dsd1314sdd1516',
    DisplayableOrderId : '1112dsd1314sdd1516',
    ShippingSpeedCategory: 'Standard',    
    DisplayableOrderDateTime :  '2017-09-12T08:03:17.255Z',
    DestName : "2A Stanlake Rd Shepherd's Bush, London W12 7HP",
    DestCountryCode : 'UK',
    DisplayableOrderComment: 'Your Package',
    LineItems : [{
        Quantity : 1,
        SellerFulfillmentOrderItemId: 'sdsd5d5sd5',
        SellerSKU : '2sdd2d2'
    }]
        }).then(({result}) => {
            var xml = result;
            parseString(xml, (err, result) => {
                if (err) {
                    console.log(err);
                }
                console.log(result);
            })
        })
            .catch((err) => {
                console.log(err);
            });

**Comment by eli8levit** [comment]
@florianbepunkt sorry for terrible view, it doesn't let me to insert the code in correct way...

**Comment by florianbepunkt** [comment]
okay, a few things… this is a bit of guesswork, but could you please try this:

```
amazonUK.FulfillmentOutboundShipment.CreateFulfillmentOrder({
  SellerFulfillmentOrderId: '1112dsd1314sdd1516',
  ShippingSpeedCategory: 'Standard',
  DisplayableOrderId: '1112dsd1314sdd1516',
  DisplayableOrderDateTime:   2017-09-12T08:03:16.50-07,
  DisplayableOrderComment: 'Your Package',
  FulfillmentPolicy: 'FillOrKill',
  FulfillmentAction: 'Hold',
  DestName: 'This ordinary guy',
  DestAddressLine1: '2A Stanlake Rd Shepherd's Bush',
  DestAddressLine2: 'Second Adress Info',
  DestAddressLine3: '',
  DestCity: 'London',
  DestPostalCode: 'W12 7HP',
  DestCountryCode: 'UK',
  LineItems:  [{
      DisplayableComment: 'lorem ipsum',
      GiftMessage: '',
      Quantity: 1, // maybe try this also as a string '1'
      SellerFulfillmentOrderItemId: 'sdsd5d5sd5',
      SellerSKU: '2sdd2d2' // please note that this sku must exists in your seller central
  }]
});
```

**Comment by eli8levit** [comment]
@florianbepunkt  it writes me "Cannot read property 'list' of undefined"
I don't have the item in inventory yet, I just want to test if it works (it doesn't).
I think that if the reason was in SellerSKU number it would throw me other error...

**Comment by florianbepunkt** [comment]
So try with a product that is in your inventory… you can kill the fulfillment order later. 

Regarding the list item error… try my pr as this could be related #68

**Comment by eli8levit** [comment]
what I have to do in order to use it?

**Comment by florianbepunkt** [comment]
Merge it… or add in the changes manually… or create a test project and install my fork
`npm i --save https://github.com/florianbepunkt/mws-api`

**Comment by eli8levit** [comment]
Ok, will try it...thank you Florian!

**Comment by eli8levit** [comment]
@florianbepunkt it seems it works, thank you very very much you saved my life!

**Comment by eli8levit** [comment]
@florianbepunkt I faced new problem, it throws an error : "Unexpected complex element termination", do you know what does it mean??

**Comment by florianbepunkt** [comment]
@JediYozh No idea to be honest. Library throws an error or you get an error back from MWS?

**Comment by eli8levit** [comment]
@florianbepunkt  the problem was in PerUnitDeclaredValue property of LineItem, it stopped to throw errors after I removed it.

**Comment by nolanamy** [comment]
@florianbepunkt @JediYozh @tyscorp Pretty sure I've got this all cleaned up (including replacing `Type.COMPLEX`) in #98. Let me know if you have feedback/issues.

**Comment by ChucklesStern** [comment]
@florianbepunkt @JediYozh 
Hi guys! hope you are well since its been a year since this post. I am also struggling with order fullfillment. Can either of you clarify the nature of the list item error? 

In my code, the ITEMS array I have my items broken down as such, which more or less conforms to the direct labelling used by AWS:

what is the best way for me to pass all my Items params into a Createfulfillment order?

amazonOrder["Items.member."+k+".FulfillmentNetworkSKU"] = encodeURI(orderArray[i].items[j].parent),
    amazonOrder["Items.member."+k+".Quantity"] = encodeURI(orderArray[i].items[j].quantity),
    amazonOrder["Items.member."+k+".SellerFulfillmentOrderItemId"] = encodeURI(orderArray[i].id),
    amazonOrder["Items.member."+k+".SellerSKU"] = encodeURI("UK-P0LV-ZDFH");

> Regarding the list item error… try my pr as this could be related #68

**Comment by florianbepunkt** [comment]
Could you pass a strack trace/error log?  And maybe the full code you are
using (credentials ommited)? Have you used my PR? We are using this in
production without problems.

Chuck Stern <notifications@github.com> schrieb am Mi., 26. Sep. 2018, 21:46:

**Comment by ChucklesStern** [comment]
HI @florianbepunkt Wow! I didnt expect such a fast response. Thank you man. Ok so I am sure I am using your PR as Ive set it to be the location of the repo in my package JSON:

 "mws-api": "git+https://github.com/florianbepunkt/mws-api.git"

the error log is: 
`2018-09-26T20:32:09.400Z	3cddf18a-c1cb-11e8-9bad-eba478487a14	TypeError: Cannot read property 'list' of undefined
    at AmazonMwsRequest.set (/var/task/grabOrderLambda/node_modules/mws-api/lib/request.js:154:22)
    at _.each (/var/task/grabOrderLambda/node_modules/mws-api/lib/request.js:178:18)
    at /var/task/grabOrderLambda/node_modules/lodash/lodash.js:4911:15
    at baseForOwn (/var/task/grabOrderLambda/node_modules/lodash/lodash.js:2996:24)
    at /var/task/grabOrderLambda/node_modules/lodash/lodash.js:4880:18
    at Function.forEach (/var/task/grabOrderLambda/node_modules/lodash/lodash.js:9344:14)
    at AmazonMwsRequest.setMultiple (/var/task/grabOrderLambda/node_modules/mws-api/lib/request.js:177:11)
    at AmazonMwsRequest.set (/var/task/grabOrderLambda/node_modules/mws-api/lib/request.js:133:25)
    at Object.CreateFulfillmentOrder (/var/task/grabOrderLambda/node_modules/mws-api/lib/api.js:37:17)
    at stripe.orders.list.then (/var/task/grabOrderLambda/grabOrders.js:195:33)
    at process._tickDomainCallback (internal/process/next_tick.js:135:7)`

and the my createfullfillment code is

`mws.FulfillmentOutboundShipment.CreateFulfillmentOrder(
  {
    "DestinationAddress.City": "San%20Francisco",
    "DestinationAddress.CountryCode": "US",
    "DestinationAddress.DistrictOrCounty": "340%20Faviola%20Street%20Unit%20814",
    "DestinationAddress.Line1": "340%20Faviola%20Street%20Unit%20814",
    "DestinationAddress.Line2": "",
    "DestinationAddress.Name": "Anton%20Jones",
    "DestinationAddress.PhoneNumber":"333333333",
    "DestinationAddress.PostalCode": "94555",
    "DestinationAddress.StateOrProvinceCode":"CA",
    "DisplayableOrderComment": "",
    "DisplayableOrderDateTime": timestamp,
    "DisplayableOrderId": "or_1D9EwNF3RbMs9n5Crr7Ie1lx",
"SellerFulfillmentOrderId": "or_1D9EwNF3RbMs9n5Crr7Ie1lx",
"ShippingSpeedCategory": "Standard",
"NotificationEmailList.member.1": "urlencodedcom",
"DestinationAddress.Name": "urlencoded",
LineItems:  [{
  
    DisplayableComment: 'lorem ipsum',
    GiftMessage: '',
    Quantity: 1, // maybe try this also as a string '1'
    SellerFulfillmentOrderItemId: 'sdsd5d5sd5',
    SellerSKU: 'UK-P0LV-ZDFH' // please note that this sku must exists in your seller central
}]});`

**Comment by ChucklesStern** [comment]
> HI @florianbepunkt Wow! I didnt expect such a fast response. Thank you man. Ok so I am sure I am using your PR as Ive set it to be the location of the repo in my package JSON:
> 
> "mws-api": "git+https://github.com/florianbepunkt/mws-api.git"
> 
> the error log is:
> `2018-09-26T20:32:09.400Z 3cddf18a-c1cb-11e8-9bad-eba478487a14 TypeError: Cannot read property 'list' of undefined at AmazonMwsRequest.set (/var/task/grabOrderLambda/node_modules/mws-api/lib/request.js:154:22) at _.each (/var/task/grabOrderLambda/node_modules/mws-api/lib/request.js:178:18) at /var/task/grabOrderLambda/node_modules/lodash/lodash.js:4911:15 at baseForOwn (/var/task/grabOrderLambda/node_modules/lodash/lodash.js:2996:24) at /var/task/grabOrderLambda/node_modules/lodash/lodash.js:4880:18 at Function.forEach (/var/task/grabOrderLambda/node_modules/lodash/lodash.js:9344:14) at AmazonMwsRequest.setMultiple (/var/task/grabOrderLambda/node_modules/mws-api/lib/request.js:177:11) at AmazonMwsRequest.set (/var/task/grabOrderLambda/node_modules/mws-api/lib/request.js:133:25) at Object.CreateFulfillmentOrder (/var/task/grabOrderLambda/node_modules/mws-api/lib/api.js:37:17) at stripe.orders.list.then (/var/task/grabOrderLambda/grabOrders.js:195:33) at process._tickDomainCallback (internal/process/next_tick.js:135:7)`
> 
> and the my createfullfillment code is
> 
> `mws.FulfillmentOutboundShipment.CreateFulfillmentOrder(
> {
> "DestinationAddress.City": "San%20Francisco",
> "DestinationAddress.CountryCode": "US",
> "DestinationAddress.DistrictOrCounty": "340%20Faviola%20Street%20Unit%20814",
> "DestinationAddress.Line1": "340%20Faviola%20Street%20Unit%20814",
> "DestinationAddress.Line2": "",
> "DestinationAddress.Name": "Anton%20Jones",
> "DestinationAddress.PhoneNumber":"333333333",
> "DestinationAddress.PostalCode": "94555",
> "DestinationAddress.StateOrProvinceCode":"CA",
> "DisplayableOrderComment": "",
> "DisplayableOrderDateTime": timestamp,
> "DisplayableOrderId": "or_1D9EwNF3RbMs9n5Crr7Ie1lx",
> "SellerFulfillmentOrderId": "or_1D9EwNF3RbMs9n5Crr7Ie1lx",
> "ShippingSpeedCategory": "Standard",
> "NotificationEmailList.member.1": "urlencodedcom",
> "DestinationAddress.Name": "urlencoded",
> LineItems: [{
> 
> ```
> DisplayableComment: 'lorem ipsum',
> GiftMessage: '',
> Quantity: 1, // maybe try this also as a string '1'
> SellerFulfillmentOrderItemId: 'sdsd5d5sd5',
> SellerSKU: 'UK-P0LV-ZDFH' // please note that this sku must exists in your seller central
> ```
> }]});`

and this is my initialization at the top of my page
`const MWSClient = require('mws-api');

const mws = new MWSClient({
  host: "mws.amazonservices.com",
  accessKeyId: 'redacted',
  secretAccessKey: secret,
  merchantId: 'redacted',
  options: {
    retry: true, // retry requests when throttled
    next: true, // auto-paginate
    limit: Infinity, // only get this number of items (NOT the same as MaxRequestsPerPage)
  }
});`

**Comment by florianbepunkt** [comment]
Okay, will check this tomorrow at the office against our code (am in
Europe/Berlin timezone). Will report back to you.

Chuck Stern <notifications@github.com> schrieb am Mi., 26. Sep. 2018, 22:43:

**Comment by ChucklesStern** [comment]
ok sounds good! Thank you so much!

**Comment by ChucklesStern** [comment]
Hi @florianbepunkt so I actually spent the entire day working on it and I just got everything working beautifully now ( I get "seller is not registered code") but thats good because I dont have my product in the warehouse yet (though FNSKUs are active).

If you dont mind I will be in touch with other questions and suggestions as I come across them. This is such a handy and amazing repo- with all PR upgrades by you I'll be relying heavily on it!
